### PR TITLE
Optimize MTrie reading single path by adding Ledger.GetSingleValue() to boost speed and reduce memory use

### LINKED ISSUE #1876: [Execution state] Optimize MTrie reading a single path
## Problem

Currently, `MTrie.UnsafeRead` reads registers in batch, so a read for a single path goes through the unnecessary steps of path partitioning and recursive calls.  

## The Proposed Solution

Create a new read function to read register for a single path.  Existing read function can use the new read function to skip unnecessary steps when `len(paths) == 1`.

`delta.View` can also be optimized by using this new read function.

### PR DESCRIPTION
**Pull request by fxamacker**
### Changes

When reading a payload from a single path, avoid needlessly path partitioning and recursive calls.  Also, optimize MTrie batch read.

- Added Ledger.GetSingleValue() to improve time/op, alloc/op, and allocs/op
- Modified delta.View.readFunc to use Ledger.GetSingleValue()
- Optimized existing MTrie batch read for single path

Closes #1876 

####  Current Ledger.Get() vs Ledger.GetSingleValue()

```
Benchstat results:
name                           old time/op    new time/op    delta
LedgerGetOneValue/batch_get-4    6.54µs ± 0%    5.24µs ± 0%  -19.92%
(p=0.000 n=9+10)

name                           old alloc/op   new alloc/op   delta
LedgerGetOneValue/batch_get-4    1.74kB ± 0%    1.62kB ± 0%   -6.85%
(p=0.000 n=10+10)

name                           old allocs/op  new allocs/op  delta
LedgerGetOneValue/batch_get-4      21.0 ± 0%      15.0 ± 0%  -28.57%
(p=0.000 n=10+10)
```

#### Optimized new Ledger.Get() vs Ledger.GetSingleValue()

```
Benchstat results:
name                           old time/op    new time/op    delta
LedgerGetOneValue/batch_get-4    5.70µs ± 0%    5.23µs ± 0%   -8.24%
(p=0.000 n=9+10)

name                           old alloc/op   new alloc/op   delta
LedgerGetOneValue/batch_get-4    1.69kB ± 0%    1.62kB ± 0%   -3.79%
(p=0.000 n=10+10)

name                           old allocs/op  new allocs/op  delta
LedgerGetOneValue/batch_get-4      18.0 ± 0%      15.0 ± 0%  -16.67%
(p=0.000 n=10+10)
```

### CODE REVIEW

**Review by ramtinms — APPROVED** [review]
Nice work, excited for this one

File: ledger/complete/mtrie/trie/trie.go [hunk 1]
```diff
@@ -195,6 +195,33 @@ func valueSizes(sizes []int, paths []ledger.Path, head *node.Node) {
 	}
 }
 
+// ReadSinglePayload reads and returns a payload for a single path.
+func (mt *MTrie) ReadSinglePayload(path ledger.Path) *ledger.Payload {
+	return readSinglePayload(path, mt.root)
```
[the remark this answers is not in the document]
Comment (fxamacker) [reply ledger/complete/mtrie/trie/trie.go:200]: Great idea!  Thanks!  👍  Done in commit d9a57fa336eb8ebc1372a3176e346713ea09a9b0.

**Comment by ramtinms** [comment]
we might actually add a single empty line to the FVM code to trigger the FVM benchmarking.

File: ledger/complete/mtrie/trie/trie.go [hunk 2]
```diff
@@ -195,6 +195,33 @@ func valueSizes(sizes []int, paths []ledger.Path, head *node.Node) {
...
+	// Traverse nodes following the path until a leaf node or nil node is reached.
+	for !head.IsLeaf() {
+		depth := ledger.NodeMaxHeight - head.Height() // distance to the tree root
+		bit := bitutils.ReadBit(pathBytes, depth)
+		if bit == 0 {
+			head = head.LeftChild()
+		} else {
+			head = head.RightChild()
+		}
+	}
```
Comment (tarakby) [inline ledger/complete/mtrie/trie/trie.go:216]: maybe sliiightly faster :) 

```suggestion
        depth := ledger.NodeMaxHeight - head.Height() // distance to the tree root
	for !head.IsLeaf() {
		bit := bitutils.ReadBit(pathBytes, depth)
		if bit == 0 {
			head = head.LeftChild()
		} else {
			head = head.RightChild()
		}
		depth++
	}
```

**Review by tarakby — APPROVED** [review]
Thanks for taking the time to write complete test 👏🏼

**Pull request merged**